File: scripts/python/parsers/object_parser.py
"""
object_parser.py
Reads Salesforce custom object metadata from SFDX source format.
Uses stdlib only (re, os, pathlib).
"""

import os
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str):
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_objects(objects_dir: str) -> dict:
    """
    Parse all custom objects under objects_dir.
    Returns {'objects': [{'apiName', 'label', 'description', 'fields': [...]}]}
    """
    objects = []
    p = Path(objects_dir)

    if not p.exists():
        return {'objects': objects}

    for obj_dir in sorted(p.iterdir()):
        if not obj_dir.is_dir():
            continue

        obj = {'apiName': obj_dir.name, 'fields': []}
        obj_file = obj_dir / f'{obj_dir.name}.object-meta.xml'

        if obj_file.exists():
            xml = obj_file.read_text(encoding='utf-8')
            obj['label'] = _extract_tag(xml, 'label')
            obj['description'] = _extract_tag(xml, 'description')
            obj['pluralLabel'] = _extract_tag(xml, 'pluralLabel')
            obj['sharingModel'] = _extract_tag(xml, 'sharingModel')

        fields_dir = obj_dir / 'fields'
        if fields_dir.exists():
            for ff in sorted(fields_dir.glob('*.field-meta.xml')):
                fxml = ff.read_text(encoding='utf-8')
                api_name = ff.name.replace('.field-meta.xml', '')
                required_val = _extract_tag(fxml, 'required')
                obj['fields'].append({
                    'apiName': api_name,
                    'type': _extract_tag(fxml, 'type'),
                    'label': _extract_tag(fxml, 'label'),
                    'required': required_val == 'true',
                    'description': _extract_tag(fxml, 'description'),
                    'length': _extract_tag(fxml, 'length'),
                })

        obj['validationRules'] = []
        vr_dir = obj_dir / 'validationRules'
        if vr_dir.exists():
            for vf in sorted(vr_dir.glob('*.validationRule-meta.xml')):
                vxml = vf.read_text(encoding='utf-8')
                obj['validationRules'].append({
                    'name':         vf.name.replace('.validationRule-meta.xml', ''),
                    'active':       _extract_tag(vxml, 'active') != 'false',
                    'errorMessage': _extract_tag(vxml, 'errorMessage'),
                    'formula':      _extract_tag(vxml, 'errorConditionFormula'),
                })

        obj['recordTypes'] = []
        rt_dir = obj_dir / 'recordTypes'
        if rt_dir.exists():
            for rf in sorted(rt_dir.glob('*.recordType-meta.xml')):
                rxml = rf.read_text(encoding='utf-8')
                obj['recordTypes'].append({
                    'name':        rf.name.replace('.recordType-meta.xml', ''),
                    'label':       _extract_tag(rxml, 'label'),
                    'active':      _extract_tag(rxml, 'active') != 'false',
                    'description': _extract_tag(rxml, 'description'),
                })

        objects.append(obj)

    return {'objects': objects}

File: scripts/node/parsers/objectParser.js
/**
 * objectParser.js
 * Reads Salesforce custom object metadata from SFDX source format.
 * Parses .object-meta.xml and .field-meta.xml files.
 */

'use strict';

const fs   = require('fs');
const path = require('path');

// Lightweight XML value extractor — avoids a heavy XML dependency
function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

function extractAll(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, 'g');
  const results = [];
  let m;
  while ((m = re.exec(xml)) !== null) results.push(m[1].trim());
  return results;
}

/**
 * @param {string} objectsDir  - path to the objects/ folder
 * @returns {{ objects: Array }}
 */
async function parseObjects(objectsDir) {
  const objects = [];

  if (!fs.existsSync(objectsDir)) return { objects };

  const entries = fs.readdirSync(objectsDir, { withFileTypes: true });

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const objDir  = path.join(objectsDir, entry.name);
    const objFile = path.join(objDir, `${entry.name}.object-meta.xml`);

    const obj = { apiName: entry.name, fields: [] };

    // Parse top-level object XML
    if (fs.existsSync(objFile)) {
      const xml         = fs.readFileSync(objFile, 'utf8');
      obj.label         = extractTag(xml, 'label');
      obj.description   = extractTag(xml, 'description');
      obj.pluralLabel   = extractTag(xml, 'pluralLabel');
      obj.sharingModel  = extractTag(xml, 'sharingModel');
    }

    // Parse individual field files
    const fieldsDir = path.join(objDir, 'fields');
    if (fs.existsSync(fieldsDir)) {
      const fieldFiles = fs.readdirSync(fieldsDir).filter(f => f.endsWith('.field-meta.xml'));
      for (const ff of fieldFiles) {
        const fxml     = fs.readFileSync(path.join(fieldsDir, ff), 'utf8');
        const apiName  = ff.replace('.field-meta.xml', '');
        obj.fields.push({
          apiName,
          type:        extractTag(fxml, 'type'),
          label:       extractTag(fxml, 'label'),
          required:    extractTag(fxml, 'required') === 'true',
          description: extractTag(fxml, 'description'),
          length:      extractTag(fxml, 'length'),
        });
      }
    }

    // Parse validation rule files
    obj.validationRules = [];
    const vrDir = path.join(objDir, 'validationRules');
    if (fs.existsSync(vrDir)) {
      for (const vf of fs.readdirSync(vrDir).filter(f => f.endsWith('.validationRule-meta.xml'))) {
        const vxml = fs.readFileSync(path.join(vrDir, vf), 'utf8');
        obj.validationRules.push({
          name:         vf.replace('.validationRule-meta.xml', ''),
          active:       extractTag(vxml, 'active') !== 'false',
          errorMessage: extractTag(vxml, 'errorMessage'),
          formula:      extractTag(vxml, 'errorConditionFormula'),
        });
      }
    }

    // Parse record type files
    obj.recordTypes = [];
    const rtDir = path.join(objDir, 'recordTypes');
    if (fs.existsSync(rtDir)) {
      for (const rf of fs.readdirSync(rtDir).filter(f => f.endsWith('.recordType-meta.xml'))) {
        const rxml = fs.readFileSync(path.join(rtDir, rf), 'utf8');
        obj.recordTypes.push({
          name:        rf.replace('.recordType-meta.xml', ''),
          label:       extractTag(rxml, 'label'),
          active:      extractTag(rxml, 'active') !== 'false',
          description: extractTag(rxml, 'description'),
        });
      }
    }

    objects.push(obj);
  }

  return { objects };
}

module.exports = { parseObjects };
